optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java

```diff
@@ -0,0 +1,201 @@
+import java.util.IdentityHashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.api.score.constraint.ConstraintMatch;
+import org.optaplanner.core.api.score.constraint.ConstraintMatchTotal;
+import org.optaplanner.core.api.score.constraint.Indictment;
+import org.optaplanner.core.impl.score.definition.ScoreDefinition;
+import org.optaplanner.core.impl.score.inliner.ScoreInliner;
+import org.optaplanner.core.impl.score.stream.ConstraintSession;
+import org.optaplanner.core.impl.score.stream.bavet.common.BavetAbstractTuple;
+import org.optaplanner.core.impl.score.stream.bavet.common.BavetNodeBuildPolicy;
+import org.optaplanner.core.impl.score.stream.bavet.common.BavetScoringNode;
+import org.optaplanner.core.impl.score.stream.bavet.common.BavetTupleState;
+import org.optaplanner.core.impl.score.stream.bavet.uni.BavetFromUniNode;
+import org.optaplanner.core.impl.score.stream.bavet.uni.BavetFromUniTuple;
+
+public final class BavetConstraintSession<Solution_> implements ConstraintSession<Solution_>  {
+
+    private final boolean constraintMatchEnabled;
+    private final Score<?> zeroScore;
+    private final ScoreInliner<?> scoreInliner;
+
+    private final Map<Class<?>, BavetFromUniNode<Object>> declaredClassToNodeMap;
+    private final int nodeOrderSize;
+    private final Map<String, BavetScoringNode> constraintIdToScoringNodeMap;
+
+    private final Map<Class<?>, List<BavetFromUniNode<Object>>> effectiveClassToNodeListMap;
+
+    private final List<Queue<BavetAbstractTuple>> nodeOrderedQueueList;
+    private final Map<Object, List<BavetFromUniTuple<Object>>> fromTupleListMap;
+
+    public BavetConstraintSession(boolean constraintMatchEnabled, ScoreDefinition scoreDefinition,
+            Map<BavetConstraint<Solution_>, Score<?>> constraintToWeightMap) {
+        this.constraintMatchEnabled = constraintMatchEnabled;
+        this.zeroScore = scoreDefinition.getZeroScore();
+        this.scoreInliner = scoreDefinition.buildScoreInliner(constraintMatchEnabled);
+        declaredClassToNodeMap = new HashMap<>(50);
+        BavetNodeBuildPolicy<Solution_> buildPolicy = new BavetNodeBuildPolicy<>(this, constraintToWeightMap.size());
+        constraintToWeightMap.forEach((constraint, constraintWeight) -> {
+            constraint.createNodes(buildPolicy, declaredClassToNodeMap, constraintWeight);
+        });
+        this.nodeOrderSize = buildPolicy.getNodeOrderMaximum() + 1;
+        constraintIdToScoringNodeMap = buildPolicy.getConstraintIdToScoringNodeMap();
+        effectiveClassToNodeListMap = new HashMap<>(declaredClassToNodeMap.size());
+        nodeOrderedQueueList = new ArrayList<>(nodeOrderSize);
+        for (int i = 0; i < nodeOrderSize; i++) {
+            nodeOrderedQueueList.add(new ArrayDeque<>(1000));
+        }
+        fromTupleListMap = new IdentityHashMap<>(1000);
```

Comment:
A thousand as a default is a bit too much for my taste. Are you sure we don't just want to let this structure grow naturally?